Add SE mode to X86Board and RiscvBoard

erin-le opened the pull request:
This PR adds SE mode to X86Board and RiscvBoard. X86DemoBoard and RiscvDemoBoard have also been modified, as the functions that were previously needed to add SE mode to them were removed.

erin-le committed:
stdlib: add SE mode to X86Board

This commit adds SE mode to X86Board. X86DemoBoard was also modified,
as functions that were previously needed to add SE mode to
X86DemoBoard were removed.

erin-le committed:
stdlib: add SE mode to RiscvBoard

This commit adds SE mode to RiscvBoard. RiscvDemoBoard has also
been modified as adding SE mode to RiscvBoard made the
overridden functions in RiscvDemoBoard obsolete.

BobbyRBruce commented on src/python/gem5/components/boards/riscv_board.py:
```diff
@@ -100,47 +101,53 @@ def __init__(
-        self._on_chip_devices = [self.platform.clint, self.platform.plic]
-        self._off_chip_devices = [self.platform.uart, self.disk, self.rng]
+        else:
+            pass
```
.Maybe just to make it very to obvious to anyone wanting to make changes, adds comment such as:

```suggestion
            # SE mode board setup.
            pass
```

BobbyRBruce commented on src/python/gem5/components/boards/riscv_board.py:
```diff
@@ -213,19 +220,31 @@ def get_dma_ports(self) -> List[Port]:
+        if self.has_io_bus():
+            return self.iobus
+        else:
+            raise NotImplementedError(
```
I don't think this exception type is the right fit. It's well worth looking up the Python docs to see what an Exception is designed to do. Here is entry for `NotImplementedError` : <https://docs.python.org/3/library/exceptions.html#NotImplementedError>

It's usage is explained as:


> In user-defined base classes, abstract methods should raise this exception when they require derived classes to override the method, or while the class is being developed to indicate that the real implementation still needs to be added.


So, this is just an exception designed to be raised in the case where someone inherits from a class but doesn't implement all the abstract methods they were supposed to.

To be honest, I've rarely found specific Exceptions in Python that really work for the kind of Exceptions we throw in gem5, so in the stdlib I've usually just used `raise Exception`, the most basic and general type. I'd recommend that.

It's been on my TODO list for a long time now to create a set of gem5 exceptions to cover this stuff (e.g., something like "InvalidGem5SimulationSetup" would be useful) but `Exception` is fine for now.

**((Do this update for all instances of this issue in this PR))**

BobbyRBruce replied:
Do this for all the other case in the PR

BobbyRBruce commented on src/python/gem5/components/boards/riscv_board.py:
```diff
@@ -213,19 +220,31 @@ def get_dma_ports(self) -> List[Port]:
+        else:
+            raise NotImplementedError(
+                "Board was not configured for FS mode and does not have an "
+                "I/O bus. Use `has_io_bus()` to check this."
```
I’d say this exception string is very specific to this board. You’ve got to remember that anyone can create a subclass of `RISCVBoard` and override `has_io_bus` to have a completely different logic. In such a situation when `get_io_bus` is called and `has_io_bus` returns `False` the exception thrown here won’t be correct.

The real, core reason this exception is raised is because we can't get the I/O bus from a board that doesn't have one. Calling this on a board without an I/O bus is simply invalid.

I’d just have the text read something like “Cannot execute `get_io_bus` : Board does not have an I/O bus to return. The board’s `has_io_bus` method can be used to check this.”

**((Make this change for other, similar exception text in this PR))**

erin-le committed:
stdlib: address requested changes to X86, Riscv boards

This commit addresses the requested changes. An additional
comment is added for clarification, the exception type is
changed, and a few of the error messages have been
modified.

Harshil2107 reviewed (approved):
lgtm
